os/fs/smartfs/smartfs_utils.c

```diff
@@ -1317,6 +1401,132 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
+#endif
+
+	ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
+	if (ret < 0) {
+		return ret;
+	}
+
+	alloc_sector = (uint16_t)ret;
+
+	/* Set the newly allocated sector's type (file or dir) */
+
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	if (sf) {
+		/* Using sector buffer and we have an open file context. Just update
+		 * the sector buffer in the open file context.
+		 */
+		memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
+		chainheader = (struct smartfs_chain_header_s *)sf->buffer;
+		chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
+		sf->bflags = SMARTFS_BFLAG_DIRTY | SMARTFS_BFLAG_NEWALLOC;
+	} else
+#endif
+	{
+		if ((type & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_DIR) {
+			chainheader_type = SMARTFS_SECTOR_TYPE_DIR;
+		} else {
+			chainheader_type = SMARTFS_SECTOR_TYPE_FILE;
+		}
+
+		smartfs_setbuffer(&readwrite, alloc_sector, offsetof(struct smartfs_chain_header_s, type), 1, (uint8_t *)&chainheader_type);
+		ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			fdbg("Error %d setting new sector type for sector %d\n", ret, alloc_sector);
+			return ret;
+		}
+	}
+
+	*sector = alloc_sector;
+	return OK;
+}
+
+/****************************************************************************
+ * Name: smartfs_invalidateentry
+ *
+ * Description: Invalidates the smartfs entry passed as a parameter.
+ *
+ * Input Parameters:
+ *  fs - pointer to smartfs mountpoint structure.
+ *  parentdirsector - parent sector of entry to be marked invalid.
+ *  offset - offset of entry in parent directory sector
+ *
+ * Returned Values:
+ *  OK - on sucessfully invalidating the entry.
+ *  ERROR - appropriate error code otherwise
+ *
+ ****************************************************************************/
+
+/* TODO IF entry is empty after invalid, we'd better release it. To do this entry needed as a parameter...*/
+int smartfs_invalidateentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, uint16_t offset)
+{
+	int ret;
+	struct smart_read_write_s readwrite;
+	uint8_t *entry_flags;
+
+	smartfs_setbuffer(&readwrite, parentdirsector, offset, sizeof(uint16_t), (uint8_t *)fs->fs_rwbuffer);
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Error %d reading sector %d\n", ret, parentdirsector);
+		return ret;
+	}
+
+	entry_flags = (uint8_t *)&fs->fs_rwbuffer[0];
+#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
+#if CONFIG_SMARTFS_ALIGNED_ACCESS
+	smartfs_wrle16(entry_flags, smartfs_rdle16(entry_flags) & ~SMARTFS_DIRENT_ACTIVE);
+#else
+	*entry_flags &= ~SMARTFS_DIRENT_ACTIVE;
+#endif
+#else
+#if CONFIG_SMARTFS_ALIGNED_ACCESS
+	smartfs_wrle16(entry_flags, smartfs_rdle16(entry_flags) | SMARTFS_DIRENT_ACTIVE);
+#else
+	*entry_flags |= SMARTFS_DIRENT_ACTIVE;
+#endif
+#endif
+	/* Now write the updated flags back to the device */
+
+	smartfs_setbuffer(&readwrite, parentdirsector, offset, sizeof(uint16_t), (uint8_t *)&fs->fs_rwbuffer[0]);
```

Comment:
If you want to use fs_rwbuffer instead of simple variable, &flag, I think you'd better use offset instead of 0.
Because we need to test this more, actually every case is not clear now, and another place using like that, we'd better align it.